File: src/astro/matching.ts
/**
 * Ashta Koota (36-guna) marriage compatibility plus Mangal Dosha.
 * Simplified classical implementation — whole-rashi Varna/Vashya, no degree-level splits.
 */
import { SIGNS, SIGN_LORDS, type Planet, type Sign } from "./signs";
import { NAK_NAMES } from "./nakshatras";
import type { Chart } from "./chart";

export type KootaName = "Varna" | "Vashya" | "Tara" | "Yoni" | "Graha Maitri" | "Gana" | "Bhakoot" | "Nadi";

export interface Koota {
  name: KootaName;
  max: number;
  points: number;
  /** Short technical note, e.g. "Groom: Deva · Bride: Manushya". */
  note: string;
  /** Each side's raw value for side-by-side display. */
  boy: string;
  girl: string;
  /** True when this koota flags a classical dosha. */
  dosha: boolean;
}

export interface Guna {
  total: number;
  max: 36;
  verdict: string;
  kootas: Koota[];
  doshas: string[];
}

// 1. Varna
const VARNA_RANK: Record<Sign, number> = {
  Cancer: 4, Scorpio: 4, Pisces: 4, Aries: 3, Leo: 3, Sagittarius: 3,
  Taurus: 2, Virgo: 2, Capricorn: 2, Gemini: 1, Libra: 1, Aquarius: 1,
};
const VARNA_NAME: Record<number, string> = { 4: "Brahmin", 3: "Kshatriya", 2: "Vaishya", 1: "Shudra" };

// 2. Vashya
const VASHYA_GROUP: Record<Sign, string> = {
  Aries: "Chatushpada", Taurus: "Chatushpada", Leo: "Vanachara", Gemini: "Manava", Virgo: "Manava",
  Libra: "Manava", Aquarius: "Manava", Cancer: "Jalachara", Pisces: "Jalachara", Capricorn: "Jalachara",
  Scorpio: "Keeta", Sagittarius: "Chatushpada",
};
const pairKey = (a: string, b: string) => [a, b].sort().join("|");
const VASHYA_SCORE = new Map<string, number>([
  [pairKey("Chatushpada", "Manava"), 1], [pairKey("Chatushpada", "Jalachara"), 1],
  [pairKey("Chatushpada", "Vanachara"), 0], [pairKey("Chatushpada", "Keeta"), 1],
  [pairKey("Manava", "Jalachara"), 1], [pairKey("Manava", "Vanachara"), 0],
  [pairKey("Manava", "Keeta"), 1], [pairKey("Jalachara", "Vanachara"), 1],
  [pairKey("Jalachara", "Keeta"), 1], [pairKey("Vanachara", "Keeta"), 1],
]);

// 4. Yoni — indexed by nakshatra
const YONI: readonly [string, "M" | "F"][] = [
  ["Horse", "M"], ["Elephant", "M"], ["Sheep", "F"], ["Serpent", "M"], ["Serpent", "F"],
  ["Dog", "F"], ["Cat", "F"], ["Sheep", "M"], ["Cat", "M"], ["Rat", "M"],
  ["Rat", "F"], ["Cow", "F"], ["Buffalo", "F"], ["Tiger", "F"], ["Buffalo", "M"],
  ["Tiger", "M"], ["Deer", "F"], ["Deer", "M"], ["Dog", "M"], ["Monkey", "F"],
  ["Mongoose", "F"], ["Monkey", "M"], ["Lion", "F"], ["Horse", "F"], ["Lion", "M"],
  ["Cow", "M"], ["Elephant", "F"],
];
const YONI_ENEMIES = new Set([
  pairKey("Horse", "Buffalo"), pairKey("Elephant", "Lion"), pairKey("Sheep", "Monkey"),
  pairKey("Dog", "Deer"), pairKey("Rat", "Cat"), pairKey("Cow", "Tiger"), pairKey("Serpent", "Mongoose"),
]);

// 5. Graha Maitri
const FRIENDS: Partial<Record<Planet, Planet[]>> = {
  Sun: ["Moon", "Mars", "Jupiter"], Moon: ["Sun", "Mercury"], Mars: ["Sun", "Moon", "Jupiter"],
  Mercury: ["Sun", "Venus"], Jupiter: ["Sun", "Moon", "Mars"], Venus: ["Mercury", "Saturn"],
  Saturn: ["Mercury", "Venus"],
};
const ENEMIES: Partial<Record<Planet, Planet[]>> = {
  Sun: ["Venus", "Saturn"], Moon: [], Mars: ["Mercury"], Mercury: ["Moon"],
  Jupiter: ["Mercury", "Venus"], Venus: ["Sun", "Moon"], Saturn: ["Sun", "Moon", "Mars"],
};
const leaning = (a: Planet, b: Planet) => (FRIENDS[a]?.includes(b) ? 1 : ENEMIES[a]?.includes(b) ? -1 : 0);
const MAITRI_POINTS: Record<number, number> = { 2: 5, 1: 4, 0: 3, [-1]: 1, [-2]: 0 };

// 6. Gana
const GANA: Record<string, "Deva" | "Manushya" | "Rakshasa"> = {
  Ashwini: "Deva", Mrigashira: "Deva", Punarvasu: "Deva", Pushya: "Deva", Hasta: "Deva", Swati: "Deva",
  Anuradha: "Deva", Shravana: "Deva", Revati: "Deva",
  Bharani: "Manushya", Rohini: "Manushya", Ardra: "Manushya", "Purva Phalguni": "Manushya",
  "Uttara Phalguni": "Manushya", "Purva Ashadha": "Manushya", "Uttara Ashadha": "Manushya",
  "Purva Bhadrapada": "Manushya", "Uttara Bhadrapada": "Manushya",
  Krittika: "Rakshasa", Ashlesha: "Rakshasa", Magha: "Rakshasa", Chitra: "Rakshasa", Vishakha: "Rakshasa",
  Jyeshtha: "Rakshasa", Mula: "Rakshasa", Dhanishta: "Rakshasa", Shatabhisha: "Rakshasa",
};
const GANA_SCORE = new Map<string, number>([
  [pairKey("Deva", "Manushya"), 5], [pairKey("Deva", "Rakshasa"), 1], [pairKey("Manushya", "Rakshasa"), 0],
]);

// 8. Nadi
const NADI_AADI = new Set(["Ashwini", "Ardra", "Punarvasu", "Uttara Phalguni", "Hasta", "Jyeshtha", "Mula", "Shatabhisha", "Purva Bhadrapada"]);
const NADI_MADHYA = new Set(["Bharani", "Mrigashira", "Pushya", "Purva Phalguni", "Chitra", "Anuradha", "Purva Ashadha", "Dhanishta", "Uttara Bhadrapada"]);
const nadiOf = (nak: string) => (NADI_AADI.has(nak) ? "Aadi" : NADI_MADHYA.has(nak) ? "Madhya" : "Antya");

const BAD_BHAKOOT_DIFFS = new Set([2, 5, 6, 8, 9, 12]);

const SCORE_INTERPRETATION: [number, string][] = [
  [32, "Excellent match — highly compatible on classical Ashta Koota grounds."],
  [24, "Good match — compatible, with a few small things to keep in mind."],
  [18, "Average match — workable, but worth discussing the concerns below with an astrologer."],
  [0, "Below the classical threshold — the score alone suggests real effort would be needed."],
];

const round2 = (n: number) => Math.round(n * 100) / 100;

export function computeGunaMilan(boy: Chart, girl: Chart): Guna {
  const bRashi = boy.rashi, gRashi = girl.rashi;
  const bNak = boy.nakshatra.name, gNak = girl.nakshatra.name;
  const bNakIdx = NAK_NAMES.indexOf(bNak), gNakIdx = NAK_NAMES.indexOf(gNak);
  const bRashiIdx = SIGNS.indexOf(bRashi), gRashiIdx = SIGNS.indexOf(gRashi);
  const kootas: Koota[] = [];

  // 1. Varna
  const bv = VARNA_RANK[bRashi], gv = VARNA_RANK[gRashi];
  kootas.push({
    name: "Varna", max: 1, points: bv >= gv ? 1 : 0, dosha: false,
    boy: VARNA_NAME[bv], girl: VARNA_NAME[gv], note: `Groom: ${VARNA_NAME[bv]} · Bride: ${VARNA_NAME[gv]}`,
  });

  // 2. Vashya
  const bg = VASHYA_GROUP[bRashi], gg = VASHYA_GROUP[gRashi];
  kootas.push({
    name: "Vashya", max: 2, points: bg === gg ? 2 : VASHYA_SCORE.get(pairKey(bg, gg)) ?? 1, dosha: false,
    boy: bg, girl: gg, note: `Groom: ${bg} · Bride: ${gg}`,
  });

  // 3. Tara
  const taraGood = (diff: number) => {
    const t = diff % 9 || 9;
    return [2, 4, 6, 8, 9].includes(t);
  };
  const diff1 = ((gNakIdx - bNakIdx + 27) % 27) + 1;
  const diff2 = ((bNakIdx - gNakIdx + 27) % 27) + 1;
  kootas.push({
    name: "Tara", max: 3, points: (taraGood(diff1) ? 1.5 : 0) + (taraGood(diff2) ? 1.5 : 0), dosha: false,
    boy: bNak, girl: gNak, note: "Birth-star counting cycle both ways",
  });

  // 4. Yoni
  const [bAnimal, bSex] = YONI[bNakIdx];
  const [gAnimal, gSex] = YONI[gNakIdx];
  const yoniPts = bAnimal === gAnimal ? (bSex === gSex ? 4 : 3) : YONI_ENEMIES.has(pairKey(bAnimal, gAnimal)) ? 0 : 2;
  kootas.push({
    name: "Yoni", max: 4, points: yoniPts, dosha: false,
    boy: bAnimal, girl: gAnimal, note: `Groom: ${bAnimal} · Bride: ${gAnimal}`,
  });

  // 5. Graha Maitri
  const bLord = SIGN_LORDS[bRashi], gLord = SIGN_LORDS[gRashi];
  const maitriPts = bLord === gLord ? 5 : MAITRI_POINTS[leaning(bLord, gLord) + leaning(gLord, bLord)];
  kootas.push({
    name: "Graha Maitri", max: 5, points: maitriPts, dosha: false,
    boy: bLord, girl: gLord, note: `Groom's Moon lord: ${bLord} · Bride's Moon lord: ${gLord}`,
  });

  // 6. Gana
  const bGana = GANA[bNak], gGana = GANA[gNak];
  kootas.push({
    name: "Gana", max: 6, points: bGana === gGana ? 6 : GANA_SCORE.get(pairKey(bGana, gGana)) ?? 6, dosha: false,
    boy: bGana, girl: gGana, note: `Groom: ${bGana} · Bride: ${gGana}`,
  });

  // 7. Bhakoot
  const bhakootDiff = ((gRashiIdx - bRashiIdx + 12) % 12) + 1;
  const bhakootDosha = BAD_BHAKOOT_DIFFS.has(bhakootDiff);
  kootas.push({
    name: "Bhakoot", max: 7, points: bhakootDosha ? 0 : 7, dosha: bhakootDosha,
    boy: bRashi, girl: gRashi,
    note: bhakootDosha ? "Bhakoot Dosha — rashi distance is inauspicious (2/12, 5/9 or 6/8)" : "Rashi distance is favourable",
  });

  // 8. Nadi
  const bNadi = nadiOf(bNak), gNadi = nadiOf(gNak);
  const nadiDosha = bNadi === gNadi;
  kootas.push({
    name: "Nadi", max: 8, points: nadiDosha ? 0 : 8, dosha: nadiDosha,
    boy: bNadi, girl: gNadi,
    note: nadiDosha ? `Nadi Dosha — both are ${bNadi} Nadi` : `Groom: ${bNadi} · Bride: ${gNadi}`,
  });

  const total = round2(kootas.reduce((s, k) => s + k.points, 0));
  const verdict = SCORE_INTERPRETATION.find(([t]) => total >= t)![1];

  const doshas: string[] = [];
  if (nadiDosha) {
    doshas.push("Nadi Dosha (same Nadi) — classically the most significant dosha; traditionally reviewed carefully as it relates to health and children.");
  }
  if (bhakootDosha) {
    doshas.push("Bhakoot Dosha (2-12 / 5-9 / 6-8 rashi distance) — can point to friction in money or emotional harmony; often softened when Graha Maitri and Gana score well.");
  }

  return { total, max: 36, verdict, kootas, doshas };
}

const MANGAL_HOUSES = new Set([1, 2, 4, 7, 8, 12]);

export interface MangalDosha {
  isManglik: boolean;
  marsHouse: number;
}

/** Basic Mangal (Manglik) Dosha: Mars in houses 1, 2, 4, 7, 8 or 12 from the Lagna. */
export function checkMangalDosha(chart: Chart): MangalDosha {
  const marsHouse = chart.houseOf.Mars;
  return { isManglik: MANGAL_HOUSES.has(marsHouse), marsHouse };
}

File: matching.py
"""
matching.py
Ashta Koota (36-guna) Vedic marriage compatibility engine, plus Mangal Dosha check.
Simplified classical implementation — whole-rashi Vashya/Varna assignment
(no degree-level Sagittarius/Capricorn splitting).
"""

from planets import SIGNS, SIGN_LORDS, NAKSHATRAS

NAK_NAMES = [n["name"] for n in NAKSHATRAS]

# ── 1. Varna (max 1) ─────────────────────────────────────────────────
VARNA_RANK = {
    "Cancer": 4, "Scorpio": 4, "Pisces": 4,               # Brahmin
    "Aries": 3, "Leo": 3, "Sagittarius": 3,                # Kshatriya
    "Taurus": 2, "Virgo": 2, "Capricorn": 2,               # Vaishya
    "Gemini": 1, "Libra": 1, "Aquarius": 1,                # Shudra
}
VARNA_NAME = {4: "Brahmin", 3: "Kshatriya", 2: "Vaishya", 1: "Shudra"}

# ── 2. Vashya (max 2) ────────────────────────────────────────────────
VASHYA_GROUP = {
    "Aries": "Chatushpada", "Taurus": "Chatushpada", "Leo": "Vanachara",
    "Gemini": "Manava", "Virgo": "Manava", "Libra": "Manava", "Aquarius": "Manava",
    "Cancer": "Jalachara", "Pisces": "Jalachara", "Capricorn": "Jalachara",
    "Scorpio": "Keeta", "Sagittarius": "Chatushpada",
}
VASHYA_SCORE = {
    frozenset(["Chatushpada"]): 2, frozenset(["Manava"]): 2,
    frozenset(["Jalachara"]): 2, frozenset(["Vanachara"]): 2, frozenset(["Keeta"]): 2,
    frozenset(["Chatushpada", "Manava"]): 1, frozenset(["Chatushpada", "Jalachara"]): 1,
    frozenset(["Chatushpada", "Vanachara"]): 0, frozenset(["Chatushpada", "Keeta"]): 1,
    frozenset(["Manava", "Jalachara"]): 1, frozenset(["Manava", "Vanachara"]): 0,
    frozenset(["Manava", "Keeta"]): 1, frozenset(["Jalachara", "Vanachara"]): 1,
    frozenset(["Jalachara", "Keeta"]): 1, frozenset(["Vanachara", "Keeta"]): 1,
}

# ── 4. Yoni (max 4) ──────────────────────────────────────────────────
YONI = [
    ("Horse", "M"), ("Elephant", "M"), ("Sheep", "F"), ("Serpent", "M"), ("Serpent", "F"),
    ("Dog", "F"), ("Cat", "F"), ("Sheep", "M"), ("Cat", "M"), ("Rat", "M"),
    ("Rat", "F"), ("Cow", "F"), ("Buffalo", "F"), ("Tiger", "F"), ("Buffalo", "M"),
    ("Tiger", "M"), ("Deer", "F"), ("Deer", "M"), ("Dog", "M"), ("Monkey", "F"),
    ("Mongoose", "F"), ("Monkey", "M"), ("Lion", "F"), ("Horse", "F"), ("Lion", "M"),
    ("Cow", "M"), ("Elephant", "F"),
]
YONI_ENEMIES = {
    frozenset(["Horse", "Buffalo"]), frozenset(["Elephant", "Lion"]),
    frozenset(["Sheep", "Monkey"]), frozenset(["Dog", "Deer"]),
    frozenset(["Rat", "Cat"]), frozenset(["Cow", "Tiger"]),
    frozenset(["Serpent", "Mongoose"]),
}

# ── 5. Graha Maitri (max 5) ──────────────────────────────────────────
FRIENDS = {
    "Sun": {"Moon", "Mars", "Jupiter"}, "Moon": {"Sun", "Mercury"},
    "Mars": {"Sun", "Moon", "Jupiter"}, "Mercury": {"Sun", "Venus"},
    "Jupiter": {"Sun", "Moon", "Mars"}, "Venus": {"Mercury", "Saturn"},
    "Saturn": {"Mercury", "Venus"},
}
ENEMIES = {
    "Sun": {"Venus", "Saturn"}, "Moon": set(),
    "Mars": {"Mercury"}, "Mercury": {"Moon"},
    "Jupiter": {"Mercury", "Venus"}, "Venus": {"Sun", "Moon"},
    "Saturn": {"Sun", "Moon", "Mars"},
}

def _leaning(a, b):
    if b in FRIENDS.get(a, set()):
        return 1
    if b in ENEMIES.get(a, set()):
        return -1
    return 0

# ── 6. Gana (max 6) ──────────────────────────────────────────────────
GANA = {
    "Ashwini": "Deva", "Mrigashira": "Deva", "Punarvasu": "Deva", "Pushya": "Deva",
    "Hasta": "Deva", "Swati": "Deva", "Anuradha": "Deva", "Shravana": "Deva", "Revati": "Deva",
    "Bharani": "Manushya", "Rohini": "Manushya", "Ardra": "Manushya",
    "Purva Phalguni": "Manushya", "Uttara Phalguni": "Manushya", "Purva Ashadha": "Manushya",
    "Uttara Ashadha": "Manushya", "Purva Bhadrapada": "Manushya", "Uttara Bhadrapada": "Manushya",
    "Krittika": "Rakshasa", "Ashlesha": "Rakshasa", "Magha": "Rakshasa", "Chitra": "Rakshasa",
    "Vishakha": "Rakshasa", "Jyeshtha": "Rakshasa", "Mula": "Rakshasa",
    "Dhanishta": "Rakshasa", "Shatabhisha": "Rakshasa",
}
GANA_SCORE = {
    frozenset(["Deva"]): 6, frozenset(["Manushya"]): 6, frozenset(["Rakshasa"]): 6,
    frozenset(["Deva", "Manushya"]): 5, frozenset(["Deva", "Rakshasa"]): 1,
    frozenset(["Manushya", "Rakshasa"]): 0,
}

# ── 8. Nadi (max 8) ──────────────────────────────────────────────────
NADI = {}
_NADI_AADI = {"Ashwini", "Ardra", "Punarvasu", "Uttara Phalguni", "Hasta", "Jyeshtha",
              "Mula", "Shatabhisha", "Purva Bhadrapada"}
_NADI_MADHYA = {"Bharani", "Mrigashira", "Pushya", "Purva Phalguni", "Chitra", "Anuradha",
                "Purva Ashadha", "Dhanishta", "Uttara Bhadrapada"}
_NADI_ANTYA = {"Krittika", "Rohini", "Ashlesha", "Magha", "Swati", "Vishakha",
               "Uttara Ashadha", "Shravana", "Revati"}
for _n in NAK_NAMES:
    if _n in _NADI_AADI:
        NADI[_n] = "Aadi"
    elif _n in _NADI_MADHYA:
        NADI[_n] = "Madhya"
    else:
        NADI[_n] = "Antya"

BAD_BHAKOOT_DIFFS = {2, 5, 6, 8, 9, 12}

SCORE_INTERPRETATION = [
    (32, "Excellent match — highly compatible on classical Ashta Koota grounds."),
    (24, "Good match — compatible, minor points to keep in mind."),
    (18, "Average match — workable, but review the doshas below with a priest/astrologer."),
    (0,  "Not recommended by classical Ashta Koota scoring alone — significant work needed."),
]


def _koota(name, max_pts, pts, note):
    return {"name": name, "max": max_pts, "points": round(pts, 2), "note": note}


def compute_guna_milan(boy_chart: dict, girl_chart: dict) -> dict:
    """
    Compute the 8-koota (Ashta Koota) compatibility score between two charts.
    Uses each chart's Moon sign (rashi) and Moon nakshatra.
    """
    b_rashi, g_rashi = boy_chart["rashi"], girl_chart["rashi"]
    b_nak, g_nak = boy_chart["nakshatra"]["name"], girl_chart["nakshatra"]["name"]
    b_nak_idx, g_nak_idx = NAK_NAMES.index(b_nak), NAK_NAMES.index(g_nak)
    b_rashi_idx, g_rashi_idx = SIGNS.index(b_rashi), SIGNS.index(g_rashi)

    kootas = []

    # 1. Varna
    bv, gv = VARNA_RANK[b_rashi], VARNA_RANK[g_rashi]
    pts = 1 if bv >= gv else 0
    kootas.append(_koota("Varna", 1, pts,
        f"Groom: {VARNA_NAME[bv]} · Bride: {VARNA_NAME[gv]}"))

    # 2. Vashya
    bg, gg = VASHYA_GROUP[b_rashi], VASHYA_GROUP[g_rashi]
    pts = VASHYA_SCORE.get(frozenset([bg, gg]), 1)
    kootas.append(_koota("Vashya", 2, pts, f"Groom: {bg} · Bride: {gg}"))

    # 3. Tara
    def tara_good(diff):
        t = diff % 9
        if t == 0:
            t = 9
        return t in (2, 4, 6, 8, 9)
    diff1 = ((g_nak_idx - b_nak_idx) % 27) + 1
    diff2 = ((b_nak_idx - g_nak_idx) % 27) + 1
    pts = (1.5 if tara_good(diff1) else 0) + (1.5 if tara_good(diff2) else 0)
    kootas.append(_koota("Tara", 3, pts, "Birth-star counting cycle both ways"))

    # 4. Yoni
    b_animal, b_sex = YONI[b_nak_idx]
    g_animal, g_sex = YONI[g_nak_idx]
    if b_animal == g_animal:
        pts = 4 if b_sex == g_sex else 3
    elif frozenset([b_animal, g_animal]) in YONI_ENEMIES:
        pts = 0
    else:
        pts = 2
    kootas.append(_koota("Yoni", 4, pts, f"Groom: {b_animal} · Bride: {g_animal}"))

    # 5. Graha Maitri
    b_lord, g_lord = SIGN_LORDS[b_rashi], SIGN_LORDS[g_rashi]
    if b_lord == g_lord:
        pts = 5
    else:
        lean = _leaning(b_lord, g_lord) + _leaning(g_lord, b_lord)
        pts = {2: 5, 1: 4, 0: 3, -1: 1, -2: 0}[lean]
    kootas.append(_koota("Graha Maitri", 5, pts, f"Groom's Moon lord: {b_lord} · Bride's Moon lord: {g_lord}"))

    # 6. Gana
    b_gana, g_gana = GANA[b_nak], GANA[g_nak]
    pts = GANA_SCORE.get(frozenset([b_gana, g_gana]), 6)
    kootas.append(_koota("Gana", 6, pts, f"Groom: {b_gana} · Bride: {g_gana}"))

    # 7. Bhakoot
    bhakoot_diff = ((g_rashi_idx - b_rashi_idx) % 12) + 1
    bhakoot_dosha = bhakoot_diff in BAD_BHAKOOT_DIFFS
    pts = 0 if bhakoot_dosha else 7
    kootas.append(_koota("Bhakoot", 7, pts,
        "Bhakoot Dosha — rashi distance is inauspicious (2/12, 5/9 or 6/8)" if bhakoot_dosha
        else "Rashi distance is favourable"))

    # 8. Nadi
    b_nadi, g_nadi = NADI[b_nak], NADI[g_nak]
    nadi_dosha = b_nadi == g_nadi
    pts = 0 if nadi_dosha else 8
    kootas.append(_koota("Nadi", 8, pts,
        f"Nadi Dosha — both are {b_nadi} Nadi" if nadi_dosha else f"Groom: {b_nadi} · Bride: {g_nadi}"))

    total = round(sum(k["points"] for k in kootas), 2)
    verdict = next(msg for threshold, msg in SCORE_INTERPRETATION if total >= threshold)

    doshas = []
    if nadi_dosha:
        doshas.append("Nadi Dosha (same Nadi) — classically considered the most significant dosha; "
                       "traditionally reviewed carefully as it relates to health/progeny compatibility.")
    if bhakoot_dosha:
        doshas.append("Bhakoot Dosha (2-12/5-9/6-8 rashi distance) — can indicate friction in "
                       "financial or emotional harmony; often mitigated if Graha Maitri and Gana score well.")

    return {
        "total": total,
        "max": 36,
        "verdict": verdict,
        "kootas": kootas,
        "doshas": doshas,
    }


MANGAL_HOUSES = {1, 2, 4, 7, 8, 12}


def check_mangal_dosha(chart: dict) -> dict:
    """Basic Mangal (Manglik) Dosha check: Mars in houses 1,2,4,7,8,12 from Lagna."""
    mars_house = chart["house_of"].get("Mars")
    is_manglik = mars_house in MANGAL_HOUSES
    return {"is_manglik": is_manglik, "mars_house": mars_house}
